browser/static/js/main_zstack.js

```diff
@@ -200,66 +165,57 @@ class Mode {
   }
 
   // keybinds that apply in bulk mode, nothing selected
-  handle_mode_none_keybind(key) {
-    if (key === 'e' && !settings.label_only) {
+  handle_mode_none_keybind(evt) {
+    if (evt.key === 'e' && !settings.label_only) {
       // toggle edit mode
-      edit_mode = !edit_mode;
+      let toggleEdit = new ToggleEdit(this);
+      actions.addFencedAction(toggleEdit);
       helper_brush_draw();
-      adjuster.preCompAdjust(state.segArray, current_highlight, edit_mode, brush, this);
-    } else if (key === 'c') {
-      // cycle forward one channel, if applicable
-      this.channel += 1;
-    } else if (key === 'C') {
-      // cycle backward one channel, if applicable
-      this.channel -= 1;
-    } else if (key === 'f') {
-      // cycle forward one feature, if applicable
-      if (feature_max > 1) {
-        this.feature = this.increment_value(this.feature, 0, feature_max - 1);
-        this.info = { feature: this.feature };
-        action('change_feature', this.info);
-        this.clear();
-      }
-    } else if (key === 'F') {
-      // cycle backward one feature, if applicable
-      if (feature_max > 1) {
-        this.feature = this.decrement_value(this.feature, 0, feature_max - 1);
-        this.info = { feature: this.feature };
-        action('change_feature', this.info);
-        this.clear();
-      }
-    } else if (key === 'p' && !rgb) {
+      adjuster.preCompAdjust(canvas.segArray, current_highlight, edit_mode, brush, this);
+    } else if (evt.key === 'c') {
+      // cycle forward one channel
+      let action = new ChangeChannel(this, adjuster, this.channel + 1);
+      actions.addFencedAction(action);
+    } else if (evt.key === 'C') {
+      // cycle backward one channel
+      let action = new ChangeChannel(this, adjuster, this.channel - 1);
+      actions.addFencedAction(action);
+    } else if (evt.key === 'f') {
+      // cycle forward one feature
+      let action = new ChangeFeature(this, this.feature + 1);
+      actions.addFencedAction(action);
+    } else if (evt.key === 'F') {
+      let action = new ChangeFeature(this, this.feature - 1);
+      actions.addFencedAction(action);
+    } else if (evt.key === 'p' && !rgb) {
       // iou cell identity prediction
       this.kind = Modes.question;
       this.action = 'predict';
       this.prompt = 'Predict cell ids for zstack? / S=PREDICT THIS FRAME / SPACE=PREDICT ALL FRAMES / ESC=CANCEL PREDICTION';
       render_info_display();
-    } else if (key === '[' && this.highlighted_cell_one !== -1) {
+    } else if (evt.key === '[' && this.highlighted_cell_one !== -1) {
       // cycle highlight to prev label
       this.highlighted_cell_one = this.decrement_value(
```

Comment:
Changing this to look like the increment code would be
```let numLabels = maxLabelsMap.get(this.feature) - 1;
      this.highlighted_cell_one = (((this.highlighted_cell_one - 2) % numLabels) + numLabels) % numLabels + 1;```
which is a little messy. 

With increment, we can easily create the cycling just with `highlighted % maxLabels + 1`, but with decrementing, we run up against the `%` remainder returning negative values & not wanting to highlight 0 as a label. I think it might be better to keep `decrement_value` around for this.

Reply:
You're right, the implementation is messier than I expected (I didn't look up `%`'s behavior in js, my bad). With that in mind I am fine keeping `decrement_value` in for now. If it had been a simple and clean change I would have liked to change it at the same time as `increment_value` but as it is, let's just consider revisiting it at some point.

Reply:
Looking back at [this resource on mod vs remainder](https://web.archive.org/web/20090717035140if_/javascript.about.com/od/problemsolving/a/modulobug.htm), I tried adding the mod function to the Number prototype, so mod becomes `x.mod(divisor)` and the code to decrement while skipping 0 becomes
```
let numLabels = maxLabelsMap.get(this.feature) - 1;
this.highlighted_cell_one = (this.highlighted_cell_one + numLabels - 2).mod(numLabels) + 1;
```
which is still a bit clumsy, but now more clearly using mod to cycle values. These lines plus a quick comment `// cycle down, skipping 0` I think is better than the decrement value function, which had three argument that a reader would have to cross reference. Adding mod to the Number prototype also cleans up some of the actions like ChangeFrame to have lines like `frame.mod(numFrames)` instead. Also, I'm changing the variables used as a divisor to be more consistent, in particular `max_frames` --> `numFrames, `channelMax` --> `numChannels`, and `feature_max` --> `numFeatures`. I'll make a separate comment about this change so it's more prominent, but here's a heads up in case you see a ton of changed lines in the javascript files with these variables.

I think better solution down the road could rework `highlighted_cell_one` to implicitly offset by 1 (so a value of 0 would highlight the first label), making the decrementing/incrementing simplify to `(highlighted_cell_one - 1).mod(numLabels)` and `(highlighted_cell_one + 1).mod(numLabels)`. This would be a more sweeping change, so I'm holding off for now.